pkg/discovery/worker/dispatcher.go

```diff
@@ -0,0 +1,96 @@
+package worker
+
+import (
+	"fmt"
+	"math"
+
+	kubeClient "k8s.io/client-go/kubernetes"
+	api "k8s.io/client-go/pkg/api/v1"
+
+	"github.com/turbonomic/kubeturbo/pkg/discovery/probe"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/task"
+
+	"github.com/golang/glog"
+)
+
+type DispatcherConfig struct {
+	kubeClient  *kubeClient.Clientset
+	probeConfig *probe.ProbeConfig
+
+	workerCount int
+}
+
+func NewDispatcherConfig(kubeClient *kubeClient.Clientset, probeConfig *probe.ProbeConfig, workerCount int) *DispatcherConfig {
+	return &DispatcherConfig{
+		kubeClient:  kubeClient,
+		probeConfig: probeConfig,
+		workerCount: workerCount,
+	}
+}
+
+type Dispatcher struct {
+	config     *DispatcherConfig
+	workerPool chan chan *task.Task
+}
+
+func NewDispatcher(config *DispatcherConfig) *Dispatcher {
+	return &Dispatcher{
+		config: config,
+		// TODO use maxWorker count for now. Improve in the future once we find a good way to get the number of task.
+		// TODO If we allow worker number burst (# of workers > maxWorker), then the extra worker would block on registering. Or we use a threshold for burst number.
+		workerPool: make(chan chan *task.Task, config.workerCount),
+	}
+}
+
+func (d *Dispatcher) Init(c *ResultCollector) {
+	for i := 0; i < d.config.workerCount; i++ {
+		workerConfig := NewK8sDiscoveryWorkerConfig(d.config.kubeClient, d.config.probeConfig.StitchingPropertyType)
+		for _, mc := range d.config.probeConfig.MonitoringConfigs {
+			workerConfig.WithMonitoringWorkerConfig(mc)
+		}
+		// create workers
+		discoveryWorker, err := NewK8sDiscoveryWorker(workerConfig)
+		if err != nil {
+			fmt.Errorf("failed to build discovery worker %s", err)
+		}
+
+		go discoveryWorker.RegisterAndRun(d, c)
+	}
+}
+
+func (d *Dispatcher) WorkerPool() chan chan *task.Task {
+	return d.workerPool
+}
+
+func (d *Dispatcher) Dispatch(nodes []*api.Node, workerCount int) int {
+	// make sure when len(node) < workerCount, worker will receive at most 1 node to discover
+	perTaskNodeLength := int(math.Ceil(float64(len(nodes)) / float64(workerCount)))
```

Comment:
(1) The design is to create a pool of workers instead of a simple task queue. Here dispatcher is aware of the number of available consumers. If we want, we can also change the code a bit to make dispatcher aware of which consumer is consumer which task. 
(2) you approach decrease the dispatcher complexity, but will increase the number of API calls, especially for those getting information shared by entities running on different nodes, like clusterID.

In general, my original thoughts are: the max number of workers is the same as num of CPU. So lets say the num of CPU is 4, and there are 16 nodes, then each worker will be responsible for 4 nodes. If we let each worker only discover 10 nodes at maximum (based on the max time of each discovery takes). Then to discover a cluster with 160 nodes, each worker will be used 4 times. Probably the discovery time is still acceptable. But, when there are 1600, then each discovery worker will be used 40 times!! Discover time can be very long. The solution is to create more temporary workers running in different pods in the cluster. We can assume at this point the overhead of creating these worker is much less than the overall discovery time.  Ideally, we should come up with an equilibrium where some tasks are finished by reused workers, some are finished by temporary workers. Of course, the number of temporary workers depends on the average time of creating a worker pod, the average time of a worker to finish a discovery and the size of the cluster.